### ISSUE TITLE
Unicode mess

### ISSUE THREAD

**Issue by EruIluvatar** [psychopy/psychopy#1525]
OS: Windows 10
Psychopy: v1.85.3 fresh install

I get a couple of Unicode{En,De}codeErrors using this version.
For example when starting the "Preferences" menu, it crashes:

> pyo version 0.8.5 (uses single precision)
> Traceback (most recent call last):
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\app\_psychopyApp.py", line 611, in showPrefs
>     prefsDlg = PreferencesDlg(app=self)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\app\preferencesDlg.py", line 159, in __init__
>     specSection=self.prefsSpec[section])
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\app\preferencesDlg.py", line 264, in makePrefPage
>     parent=panel, name=pLabel, value=thisPref, spec=thisSpec)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\app\preferencesDlg.py", line 375, in __init__
>     devs = sound.getDevices('output')
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 68, in getDevices
>     devName = dev['name'].decode(osEncoding) # convert to unicode
>   File "C:\Program Files (x86)\PsychoPy2\lib\encodings\mbcs.py", line 21, in decode
>     return mbcs_decode(input, errors, True)
> UnicodeEncodeError: 'ascii' codec can't encode character u'\xe4' in position 48: ordinal not in range(128)

Similarly the demo "soundStimuli.py" crashes:

> pyo version 0.8.5 (uses single precision)
> Using pyo(with None) for sounds
> Traceback (most recent call last):
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\demos\coder\stimuli\soundStimuli.py", line 25, in <module>
>     highA = sound.Sound('A', octave=3, sampleRate=44100, secs=0.8, stereo=True)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 279, in __init__
>     init(rate=sampleRate)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 128, in init
>     audioDriver, outputID = _bestDriver(devNames, devIDs)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 26, in _bestDriver
>     preferredDrivers = prefs.general['audioDriver']
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\configobj.py", line 563, in __getitem__
>     new = [_check(entry) for entry in val]
> UnicodeDecodeError: 'utf8' codec can't decode byte 0xe4 in position 48: invalid continuation byte

I need sounds to work but creating a "< class Sound ... >" object crashes everytime.
What is wrong with this install?

**Comment by peircej** [comment]
OK, thanks for raising this. Basically, your audio card appears to have a unicode character (with an umlaut?) in its driver name. It's hard for me to debug this directly because it's dependent on your system and the language encoding. Would you be able to help me work out a fix?

To debug the first error it would help if you could paste this into a coder script and paste the output back here (select the output and right-click to get the copy option)
```python
import pyo
import sys

print(sys.getfilesystemencoding())
inputs, outputs = pyo.pa_get_devices_infos()
for this in outputs:
    dev = outputs[this]
    print(repr(dev['name']))
```

For the second issue could you send me your user preferences file from HOME/AppData/Roaming/psychopy2/userprefs.cfg
You should be able to add it here as an attachment.

cheers,
Jon

**Comment by EruIluvatar** [comment]
Alright, Umlaut was a good guess. It's called "High Definition Audio-Gerät".

> pyo version 0.8.5 (uses single precision)
> mbcs
> u'Microsoft Soundmapper - Output'
> u'2460G4 (Intel(R) Display-Audio)'
> u'Digitalaudio (S/PDIF) (High Def'
> u'???'
> u'2460G4 (Intel(R) Display-Audio)'
> u'???'
> u'Digitalaudio (S/PDIF) (High Definition Audio-Ger\xe4t)'
> u'2460G4 (Intel(R) Display-Audio)'
> u'SPDIF Out (HD Audio-SPDIF-Ausgabe)'
> u'Output (Intel(R) Display-Audio - Ausgang 3)'
> Traceback (most recent call last):
>   File "C:\Users\stbo\Downloads\audio-test.py", line 6, in <module>
>     for this in outputs:
> UnicodeDecodeError: 'utf8' codec can't decode byte 0xe4 in position 48: invalid continuation byte

See attachment for the userprefs.cfg (with .txt extension, github doesn't allow .cfg).
[userPrefs.txt](https://github.com/psychopy/psychopy/files/1323748/userPrefs.txt)

**Comment by peircej** [comment]
OK, let's try this. Could you edit the file `C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py`?

so that the line (#68 or so):
```
       devName = dev['name'].decode(osEncoding) # convert to unicode
```
becomes the lines:
```
        try:  # convert to unicode
            devName = dev['name'].decode(osEncoding)
        except UnicodeEncodeError:  # if that fails try the current encoding
            devName = dev['name']
```

and see if the errors go away? I have a feeling what's happening is that the name of the device had already been "decoded" by pyo. If hits works for you then what's still a puzzle is "why this was not the case on other machine, for which the decode line was required!"

**Comment by hsogo** [comment]
Hi,

I modified Jon's code and run it on my Japanese Windows 7 PC.
The code is following:

```
import pyo
import sys

print(sys.getfilesystemencoding())
inputs, outputs = pyo.pa_get_devices_infos()
print(len(outputs))
for this in outputs:
    dev = outputs[this]
    print(type(dev['name']))
    print(repr(dev['name']))
```

When I run this code on PsychoPy 1.82.02 (which includes pyo 0.6.6), I got following output.
Four devices were found and type of dev['name'] was str.

> pyo version 0.6.6 (uses single precision)
> mbcs
> 4
> <type 'str'>
> 'Microsoft \x83T\x83E\x83\x93\x83h \x83}\x83b\x83p\x81[ - Output'
> <type 'str'>
> '\x83X\x83s\x81[\x83J\x81[ (2- High Definition '
> <type 'str'>
> '\x83v\x83\x89\x83C\x83}\x83\x8a \x83T\x83E\x83\x93\x83h \x83h\x83\x89\x83C\x83o\x81['
> <type 'str'>
> '\x83X\x83s\x81[\x83J\x81[ (2- High Definition Audio \x83f\x83o\x83C\x83X)'

On the other hand, when I run the code on PsychoPy 1.83.04 (which includes pyo 0.8.5) **on the same machine**, I got difference output.

> pyo version 0.8.5 (uses single precision)
> mbcs
> 6
> <type 'unicode'>
> u'???'
> <type 'unicode'>
> u'???'
> <type 'unicode'>
> u'???'
> <type 'unicode'>
> u'???'
> <type 'unicode'>
> u'\u30b9\u30d4\u30fc\u30ab\u30fc (2- High Definition Audio \u30c7\u30d0\u30a4\u30b9)'
> <type 'unicode'>
> u'Speakers (HD Audio \u30d8\u30c3\u30c9\u30d5\u30a9\u30f3/\u30b9\u30d4\u30fc\u30ab\u30fc)'
> Traceback (most recent call last):
>   File "C:\Users\hsogo\Desktop\pyo_test.py", line 7, in <module>
>     for this in outputs:
> UnicodeDecodeError: 'utf8' codec can't decode byte 0x83 in position 0: invalid start byte

Firstly, 6 devices were found instead of 4.  Secondly, type of dev['name'] was unicode.
I guess that this issue is caused by changes in the way pyo.pa_get_devices_infos() works.

**Comment by peircej** [comment]
Thanks @hsogo. That's very informative. I did suspect that this must be a change in pyo (and/or in portaudio underneath that).
I uploaded a fix that should work on both systems here: https://github.com/psychopy/psychopy/commit/196abd9abc75cf60b66a68297ec02b6082a82dc4

Given your info we *could* change that code to be:
```
       if type(dev['name']) == bytes:  # convert to unicode for pyo<0.8.5
            devName = dev['name'].decode(osEncoding)
       else:  # has already been converted to unicode
            devName = dev['name']
       devs[devName] = dev
```
on the basis that `if...else` is a little nicer than `try...except`

**Comment by EruIluvatar** [comment]
I can confirm your suggested fix works for the first problem: The preferences window opens now without complaint.
soundStimuli.py is still unresolved.

**Comment by EruIluvatar** [comment]
I don't get where the second crash is coming from.
When I debugging `_bestDriver(devNames, devIDs)` and evaluate the following in PyCharm before executing the code `preferredDrivers = prefs.general['audioDriver']` (line 26) I get the expected values from userPrefs.cfg and no crash:

> <type 'list'>: [u'Primary Sound', u'ASIO', u'Audigy']

**Comment by peircej** [comment]
Yes, I'm also thrown by this one. 
But you're sure it's still giving the error message above?

**Comment by hsogo** [comment]
I tried Standalone PsychoPy 1.85.3b on my Japanese Windows10 machine and got the same error.

```
from psychopy import prefs
from psychopy import sound

print prefs.general['audioDriver'] 
highA = sound.Sound('A', octave=3, sampleRate=44100, secs=0.8, stereo=True)
```

Output of this code is:

> pyo version 0.8.5 (uses single precision)
> [u'Primary Sound', u'ASIO', u'Audigy']
> Traceback (most recent call last):
>   File "C:\Users\Hiroyuki\Desktop\test.py", line 5, in <module>
>     highA = sound.Sound('A', octave=3, sampleRate=44100, secs=0.8, stereo=True)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 282, in \_\_init\_\_
>     init(rate=sampleRate)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 131, in init
>     audioDriver, outputID = _bestDriver(devNames, devIDs)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 26, in _bestDriver
>     preferredDrivers = prefs.general['audioDriver']
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\configobj.py", line 563, in \_\_getitem\_\_
>     new = [_check(entry) for entry in val]
> UnicodeDecodeError: 'utf8' codec can't decode byte 0x83 in position 0: invalid start byte

It seems that value of prefs.general['audioDriver'] would be updated between call of sound.Sound() and _bestDriver().

**Comment by hsogo** [comment]
I confirmed that pyo.pa_get_output_devices() in backend_pyo.init() (line 131) affects prefs.general['audioDriver']. Following code it the simplest to raise this error, I think.

```
from psychopy import prefs
import pyo

print prefs.general['audioDriver'] # successful
pyo.pa_get_output_devices()
print prefs.general['audioDriver'] # fail
```

> pyo version 0.8.5 (uses single precision)
> [u'Primary Sound', u'ASIO', u'Audigy']
> Traceback (most recent call last):
>   File "C:\Users\Hiroyuki\Desktop\test.py", line 6, in <module>
>     print prefs.general['audioDriver']
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\configobj.py", line 563, in __getitem__
>     new = [_check(entry) for entry in val]
> UnicodeDecodeError: 'utf8' codec can't decode byte 0x83 in position 0: invalid start byte

I'm sorry I have no idea what is happening.

**Comment by EruIluvatar** [comment]
> But you're sure it's still giving the error message above?

Yes it's the same except for the line numbers in backend_pyo.py of course.

> pyo version 0.8.5 (uses single precision)
> Using pyo(with None) for sounds
> Traceback (most recent call last):
>   File "C:\Users\stbo\Downloads\soundStimuli.py", line 25, in <module>
>     highA = sound.Sound('A', octave=3, sampleRate=44100, secs=0.8, stereo=True)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 282, in __init__
>     init(rate=sampleRate)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 131, in init
>     audioDriver, outputID = _bestDriver(devNames, devIDs)
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 26, in _bestDriver
>     preferredDrivers = prefs.general['audioDriver']
>   File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\configobj.py", line 563, in __getitem__
>     new = [_check(entry) for entry in val]
> UnicodeDecodeError: 'utf8' codec can't decode byte 0xe4 in position 48: invalid continuation byte

**Comment by hsogo** [comment]
I found following code is enough to raise an error! (tested with Standalone PsychoPy 1.85.3b on Japanese 32bit Windows 7)

```
import pyo
pyo.pa_get_output_devices()
```

> pyo version 0.8.5 (uses single precision)
> Exception UnicodeDecodeError: UnicodeDecodeError('utf8', '\x83v\x83\x89\x83C\x83}\x83\x8a \x83T\x83E\x83\x93\x83h \x83h\x83\x89\x83C\x83o\x81[', 0, 1, 'invalid start byte') in <module 'threading' from 'C:\Program Files\PsychoPy2\lib\threading.pyc'> ignored

Interestingly, the message says that error occurs in threading.pyc.
This error cannot be caught by  try-except sentence because it occurs in different thread.

```
import pyo
try:
    pyo.pa_get_output_devices()
except:
    print("error!")
print("done")
```

Output is following. Note that "done" is printed and then UnicodeDecodeError occurs.

> pyo version 0.8.5 (uses single precision)
> done
> Exception UnicodeDecodeError: UnicodeDecodeError('utf8', '\x83v\x83\x89\x83C\x83}\x83\x8a \x83T\x83E\x83\x93\x83h \x83h\x83\x89\x83C\x83o\x81[', 0, 1, 'invalid start byte') in <module 'threading' from 'C:\Program Files\PsychoPy2\lib\threading.pyc'> ignored

Inserting time.sleep(1) before print(), UnicodeDecodeError occurs before executing print() and the script is halted.

```
import pyo
import time
try:
    pyo.pa_get_output_devices()
except:
    print("error!")
time.sleep(1)
print("done")
```

Output is following. The name of threading.pyc is not appeared in the error message.
This error message is the same as that occurs in PsychoPy's backend_pyo.py on my PC.

> pyo version 0.8.5 (uses single precision)
> Traceback (most recent call last):
>   File "C:\Users\psyuser\Desktop\untitled2.py", line 7, in <module>
>     time.sleep(1)
> UnicodeDecodeError: 'utf8' codec can't decode byte 0x83 in position 0: invalid start byte

Including time.sleep(1) in "try" section, the error can be caught by try-except sentence.

```
import pyo
import time
try:
    pyo.pa_get_output_devices()
    time.sleep(1)
except:
    print("error!")
print("done")
```

> pyo version 0.8.5 (uses single precision)
> error!
> done

In conclusion, this error occurs in pyo.  To catch this error by try-except sentence, we have to wait until the error is raised.

**Comment by EruIluvatar** [comment]
The problem is two-fold I would say:

1. the pyo library uses multiple threads, so we probably need some synchronization.
See: [https://github.com/belangeo/pyo/blob/14692dd942772ba5bae73146625e19ab2b1d59c2/src/engine/ad_portaudio.c#L606](https://github.com/belangeo/pyo/blob/14692dd942772ba5bae73146625e19ab2b1d59c2/src/engine/ad_portaudio.c#L606)
There's more information the topic:
[https://docs.python.org/2.0/api/threads.html](https://docs.python.org/2.0/api/threads.html)
2. pyo probably messes up the char encoding (?!)

**Comment by peircej** [comment]
@EruIluvatar I don't think the question of threads is relevant here. Yes, we need to try and minimise sound latencies but the presence or absence of a thread is only one part of that.

@hsogo I note that pyo now has a version 0.8.7. Would you be able to see if it's fixed there in a local Python installation? If so then I'll try to rebuild a package for that.

You're both using windows, right? No signs of this being a problem on macs at the moment?

thanks

**Comment by EruIluvatar** [comment]
Yes Windows 10 here, no way to test on a Mac at the moment.

I just tried pyo 0.8.7 (copied all the contents from site-packages to PsychoPy's site-packages folder), doesn't resolve the problem unfortunately.

**Comment by EruIluvatar** [comment]
I just tested the snippets on linux with psychopy 1.85.3-1 (with python2-pyo 0.8.7, [details](https://aur.archlinux.org/packages/psychopy/)), also german locale (de_CH.UTF-8).

```
import pyo
import sys

print(sys.getfilesystemencoding())
inputs, outputs = pyo.pa_get_devices_infos()
for this in outputs:
    dev = outputs[this]
    print(repr(dev['name']))
```
No Umlaute present as you can see:

> pyo version 0.8.7 (uses single precision)
> /usr/lib/python2.7/site-packages/wx-3.0-gtk2/wx/_core.py:16629: UserWarning: wxPython/wxWidgets release number mismatch
>   warnings.warn("wxPython/wxWidgets release number mismatch")
> UTF-8
> u'HDA Intel PCH: ALC293 Analog (hw:0,0)'
> u'HDA Intel PCH: HDMI 0 (hw:0,3)'
> u'HDA Intel PCH: HDMI 1 (hw:0,7)'
> u'HDA Intel PCH: HDMI 2 (hw:0,8)'
> u'HDA Intel PCH: HDMI 3 (hw:0,9)'
> u'HDA Intel PCH: HDMI 4 (hw:0,10)'
> u'sysdefault'
> u'front'
> u'surround40'
> u'surround51'
> u'surround71'
> u'hdmi'
> u'pulse'
> u'dmix'
> u'default'
> ALSA lib pcm.c:2501:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
> ALSA lib pcm.c:2501:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
> ALSA lib pcm.c:2501:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
> ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
> ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
> ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
> ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
> connect(2) call to /dev/shm/jack-1000/default/jack_0 failed (err=No such file or directory)
> attempt to connect to server failed

---
```
from psychopy import prefs
import pyo

print prefs.general['audioDriver']
pyo.pa_get_output_devices()
print prefs.general['audioDriver']
```
Seems like the encoding issue doesn't exist here (again: no umlaute).
(It somehow got the audio service wrong since I use pulseaudio, but that's another story.)

> pyo version 0.8.7 (uses single precision)
> [u'portaudio']
> /usr/lib/python2.7/site-packages/wx-3.0-gtk2/wx/_core.py:16629: UserWarning: wxPython/wxWidgets release number mismatch
>   warnings.warn("wxPython/wxWidgets release number mismatch")
> [u'portaudio']
> ALSA lib pcm.c:2501:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
> ALSA lib pcm.c:2501:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
> ALSA lib pcm.c:2501:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
> ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
> ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
> ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
> ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
> connect(2) call to /dev/shm/jack-1000/default/jack_0 failed (err=No such file or directory)
> attempt to connect to server failed

**Comment by hsogo** [comment]
I use 64bit Windows 10 and 32bit Windows 7.
I tried pyo 0.8.0 to 0.8.7 with Standalone PsychoPy 1.85.3b for Windows. Only 0.8.0 worked fine.
(Standalone PsychoPy 1.84.2 comes with pyo 0.8.0, which also works fine.)

From pyo's changelog (`https://github.com/belangeo/pyo/blob/master/ChangeLog`), Python 3 migration has done between 0.8.0 and 0.8.1 (2016-11-06).  I guess that this change may cause error on Python 2. 

Do we need new features introduced to pyo later than 0.8.0?
If not, bundling pyo 0.8.0 with Windows Standalone PsychoPy until this problem is fixed might be a safe option.

**Comment by EruIluvatar** [comment]
FWIW, pyo 0.8.0 does the trick here too.
It's only a temporary solution (as we have to copy the pyo files in every laboratory install by hand) but it's good enough for now.
We probably should report this upstream?!

**Comment by peircej** [comment]
We should certainly report the issue to the pyo team but, in all honesty I'd like simply to move away from it! `pyo` has frequently included bugs like this that either prevent it from running on one of the platforms or has prevented its thread from closing properly. I'm keen to push the switch to sounddevice instead; it's simpler and the bugs I've encountered there have never been deal-breakers.

**Comment by hoechenberger** [comment]
Potentially related: #1284

**Comment by Losses** [comment]
This is a serious problem, I even can't open the setting panel to change the audio lib to `pygame` to avoid this problem, consider to rollback to an older version until this problem is fixed.

the console prints the following information: 
```
Welcome to PsychoPy2!
v1.85.3
pyo version 0.8.5 (uses single precision)
Traceback (most recent call last):
  File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\app\_psychopyApp.py", line 611, in showPrefs
    prefsDlg = PreferencesDlg(app=self)
  File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\app\preferencesDlg.py", line 159, in __init__
    specSection=self.prefsSpec[section])
  File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\app\preferencesDlg.py", line 264, in makePrefPage
    parent=panel, name=pLabel, value=thisPref, spec=thisSpec)
  File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\app\preferencesDlg.py", line 375, in __init__
    devs = sound.getDevices('output')
  File "C:\Program Files (x86)\PsychoPy2\lib\site-packages\psychopy\sound\backend_pyo.py", line 68, in getDevices
    devName = dev['name'].decode(osEncoding) # convert to unicode
  File "C:\Program Files (x86)\PsychoPy2\lib\encodings\mbcs.py", line 21, in decode
    return mbcs_decode(input, errors, True)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-2: ordinal not in range(128)
```

**Comment by hsogo** [comment]
Unicode error in pa_get_devices_infos() seems to be fixed in the latest pyo.

[Fixed pa_get_devices_infos() function on Windows.](https://github.com/belangeo/pyo/commit/8f78ecdebcf52fb7806d44bbeb982a237625236b)

**Comment by peircej** [comment]
As far as I can tell the fix is not part of the 0.8.8 release so presumably we need to wait for the next one

**Comment by hsogo** [comment]
Indeed.  We should avoid using `pyo.pa_get_devices_infos()` until new version of pyo is released and tested.

By the way, I'm troubled in pyo's runtime error in some Windows PCs.  Pyo seems to have bug(s) in memory management and sometimes causes runtime error when closing python process.  I haven't specified the exact condition for runtime error, but the error frequently happens after using PsychoPy's preference dialog and configuration/benchmark wizard.  As for the preference dialog, the error can be avoided by setting audio library other than pyo.  However, as for the configuration/benchmark wizard, pyo is imported regardless of audio library's setting and runtime error happens at the end of the wizard.  I'm afraid that this error would make first-time users anxious. (Note: pyo doesn't always cause runtime error; for example, coder's "soundStimuli" demo works fine on the same machine.  In addition, I haven't experienced this issue in Ubuntu).

We can't get rid of pyo right now because microphone and voiceKey depends on pyo, but is it necessary to test whether pyo can be imported in  the configuration/benchmark wizard?  If not, what do you think about removing pyo installation test from the the configuration/benchmark wizard?

**Comment by peircej** [comment]
Agreed, pyo has caused us quite a few issues down the years with memory errors and not closing itself down etc.
I think sounddevice is promising as an alternative, due to its simplicity, but I don't have time to rewrite mic for that just yet.

**Comment by peircej** [comment]
We could certainly remove it from the benchmark though - that doesn't seem useful as we can't actually test its performance.

**Comment by hsogo** [comment]
I tried testing availability of pyo without importing it.

[BF: check availability of pyo without importing it](https://github.com/hsogo/psychopy/commit/d3313395da904b9ecf6ea181f71d1df52a953016#diff-79eafd7bcbcddccdbf6b83018bf91fb2)

This works without error in my systems in which `import pyo` causes crash.
Downsides is that we can't know version number.

**Comment by hsogo** [comment]
@peircej 
Pyo 0.8.9 was released [(http://ajaxsoundstudio.com/software/pyo/)](http://ajaxsoundstudio.com/software/pyo/) and I tried it.  Unicode errors don't occur in 0.8.9, but sound device names are still corrupted (as reported [here](https://discourse.psychopy.org/t/python3-compatibility-issues/3581/12)).  More importantly, runtime error at terminating python process in MS Windows is not fixed.  So, we should avoid importing pyo (and psychopy.sound.backend_pyo) unless it is truly necessary.

One of the most annoying problem is that Preference Dialog imports psychopy.sound to get audio device list.  This causes runtime error after opening Preference Dialog.  I think that this task can be done with sounddevice.  Here is my attempt.

[TESTS: use sounddevice to get audio device list in Preference Dialo](https://github.com/hsogo/psychopy/commit/3ec32faed8d27037d5f9effdc5980a59007c8cdb)

What do you think?  If it is OK, I'll merge this to my main branch and send a pull request.

**Comment by hsogo** [comment]
Jon, thank you for your reply to my commit.  I've sent pull request #1720.

**Comment by TEParsons** [comment]
Closing this as `pyo` is no longer within the core lib (we use sounddevice and psychtoolbox now)
